src/components/Trade/hooks/useSwapper/useSwapper.tsx

```diff
@@ -151,6 +166,76 @@ export const useSwapper = () => {
     ],
   )
 
+  const checkApprovalNeeded = useCallback(async () => {
+    const adapterManager = getChainAdapterManager()
+    const adapter = adapterManager.get(sellAsset.chainId)
+
+    if (!adapter) throw Error(`no chain adapter found for chain Id: ${sellAsset.chainId}`)
+    if (!wallet) throw new Error('no wallet available')
+    if (!activeQuote) throw new Error('no activeQuote available')
+
+    // No approval needed for selling a fee asset
+    if (sellAsset.assetId === adapter.getFeeAssetId()) {
+      return false
+    }
+
+    const ownerAddress = await adapter.getAddress({
+      wallet,
+      accountNumber: activeQuote.accountNumber,
+    })
+
+    const { assetReference: sellAssetErc20Address } = fromAssetId(sellAsset.assetId)
+    const web3 = getWeb3InstanceByChainId(sellAsset.chainId)
+
+    const allowanceResult = await getERC20Allowance({
+      web3,
+      erc20AllowanceAbi,
+      sellAssetErc20Address,
+      spenderAddress: activeQuote.allowanceContract,
+      ownerAddress,
+    })
+    const allowanceOnChain = bnOrZero(allowanceResult)
+
+    return allowanceOnChain.lt(bnOrZero(activeQuote.sellAmountBeforeFeesCryptoBaseUnit))
+  }, [activeQuote, sellAsset.assetId, sellAsset.chainId, wallet])
+
+  const createBuildApprovalTxInput = useCallback(
+    (isExactAllowance: boolean): Promise<evm.BuildCustomTxInput> => {
+      const adapterManager = getChainAdapterManager()
+      const adapter = adapterManager.get(sellAsset.chainId)
+
+      if (!activeQuote) throw new Error('no activeQuote available')
+      if (!wallet) throw new Error('no wallet available')
+      if (!adapter || !isEvmChainAdapter(adapter))
+        throw Error(`no valid EVM chain adapter found for chain Id: ${sellAsset.chainId}`)
+
+      const approvalAmountCryptoBaseUnit = isExactAllowance
+        ? activeQuote.sellAmountBeforeFeesCryptoBaseUnit
+        : MAX_ALLOWANCE
+
+      const web3 = getWeb3InstanceByChainId(sellAsset.chainId)
+
+      const { assetReference: erc20ContractAddress } = fromAssetId(sellAsset.assetId)
+
+      const erc20ContractData = getApproveContractData({
+        approvalAmountCryptoBaseUnit,
+        spenderAddress: activeQuote.allowanceContract,
+        erc20ContractAddress,
+        web3,
+      })
+
+      return createBuildCustomTxInput({
+        accountNumber: activeQuote.accountNumber,
+        adapter,
+        erc20ContractAddress,
```

Comment:
these names as stated elsewhere are the ones that I would just make generic (to, data, value) since it is just a generic tx helper for any custom tx you want to build